app/src/main/java/org/mozilla/focus/utils/IntentUtils.kt

```diff
@@ -3,142 +3,68 @@
-                val externalAppTitle = info?.loadLabel(packageManager) ?: "(null)"
+    fun getIntentChooser(
+        context: Context,
+        intent: Intent,
+        chooserTitle: CharSequence? = null
+    ): Intent? {
+        val chooserIntent: Intent?
+        val resolveInfos = context.packageManager.queryIntentActivities(intent, 0).toHashSet()
+
+        val excludedComponentNames = HashSet<ComponentName>()
+        resolveInfos.forEach {
+            val activityInfo = it.activityInfo
+            if (activityInfo.packageName == context.packageName) {
+                val componentName = ComponentName(activityInfo.packageName, activityInfo.name)
+                excludedComponentNames.add(componentName)
+            }
+        }
 
-                showConfirmationDialog(
-                    context,
-                    intent,
-                    context.getString(R.string.external_app_prompt_title),
-                    R.string.external_app_prompt,
-                    externalAppTitle
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            chooserIntent = Intent.createChooser(intent, chooserTitle)
+                .putExtra(Intent.EXTRA_EXCLUDE_COMPONENTS, excludedComponentNames.toTypedArray())
+        } else {
+            val targetIntents: MutableList<Intent> = ArrayList()
+            for (resolveInfo in resolveInfos) {
+                val activityInfo = resolveInfo.activityInfo
+                if (excludedComponentNames.contains(ComponentName(activityInfo.packageName, activityInfo.name)))
+                    continue
+                val targetIntent = Intent(intent)
+                targetIntent.setPackage(activityInfo.packageName)
+                targetIntent.component =
+                    ComponentName(activityInfo.packageName, activityInfo.name)
+                // wrap with LabeledIntent to show correct name and icon
+                val labeledIntent = LabeledIntent(
+                    targetIntent,
+                    activityInfo.packageName,
+                    resolveInfo.labelRes,
+                    resolveInfo.icon
                 )
+                // add filtered intent to a list
+                targetIntents.add(labeledIntent)
             }
-            else -> {
-                val chooserTitle = context.getString(R.string.external_multiple_apps_matched_exit)
-                val chooserIntent = Intent.createChooser(intent, chooserTitle)
-                context.startActivity(chooserIntent)
+            // deal with M list seperate problem
+            chooserIntent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                // create chooser with empty intent in M could fix the empty cells problem
+                Intent.createChooser(Intent(), chooserTitle)
+            } else {
+                // create chooser with one target intent below M
+                Intent.createChooser(targetIntents.removeAt(0), chooserTitle)
             }
```

Comment:
Offering more context about these issues and the need for separate approaches would be nice.